slam_toolbox/solvers/ceres_solver.cpp

```diff
@@ -232,6 +233,31 @@ const karto::ScanSolver::IdPoseVector& CeresSolver::GetCorrections() const
   return corrections_;
 }
 
+/*****************************************************************************/
+Eigen::SparseMatrix<double> CeresSolver::GetInformationMatrix(
+    std::unordered_map<int, Eigen::Index> * ordering) const
+/****************************************************************************/
+{
+  if (ordering) {
+    Eigen::Index index = 0u;
+    std::vector<double*> parameter_blocks;
+    problem_->GetParameterBlocks(&parameter_blocks);
+    for (auto * block : parameter_blocks) {
+      (*ordering)[(*nodes_inverted_)[block]] = index;
```

Comment:
Please explain this

Reply:
TL; DR I need a mechanism to recover the order of each node variable to make sense out of the information matrix.

When computing gradients and jacobians, Ceres matches the parameter block order (see [here](http://ceres-solver.org/nnls_modeling.html#_CPPv4N5ceres7Problem15EvaluateOptions16parameter_blocksE)). Parameter block order matches insertion order, until you [remove a block](https://github.com/ceres-solver/ceres-solver/blob/c8658c8992acd1a862c6a193b3f21e1ee0274344/internal/ceres/problem_impl.cc#L414-L439). Then that ordering is lost. Completely. This bit of code reconstructs that ordering, using a hashmap to recover the node ID for a given block.

Reply:
Is it not possible that we can combine the 2 maps with a struct so we have 1 data structure containing all of this information? I don't love the idea of having these 2 independent tables, it doesn't seem good

Reply:
Hmm, what do you mean exactly? To build a bi-directional map of sorts backed by two hash maps? Or to have an actual bi-directional map?

Reply:
Well, using a proper bidirectional map was a _tiny_ bit more troublesome than anticipated, as what I was doing here wasn't quite it (i.e. `Eigen::Vector3d` isn't `double *`). See https://github.com/SteveMacenski/slam_toolbox/pull/468/commits/1041a52eae3d3c75d1c48c87dedb3cae8bffb3d5. On the flip side, using a pool for parameter blocks may turn out to be beneficial to cache performance.

Reply:
Why not create a struct as the value in the map that contains more information than just the single entry?

Reply:
> Why not create a struct as the value in the map that contains more information than just the single entry?

Hmm, I don't follow. The original hash map would map node ID to parameter block. I need to perform the reverse query. My previous approach used two hash maps. The current one uses a bidirectional map. Are you proposing to have a single hash map and perform a linear search every time we need to map a parameter block to a node ID? That's O(n (n + 1) / 2) in the worst case (i.e. next parameter block always found last).